### PR TITLE
Filter rule scope in SymbolicExecutionFactory

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
It does not change the behavior for now, because all rules run only on the main code. 

But having at least one rule running on Test&Main or Test causes `SymbolicExecutionRunner` to contain at least one diagnostic descriptor that should execute so all rules are executed, causing FPs in terms of their scope.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs [hunk 1]
```diff
@@ -82,25 +82,27 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                Analyze(analysisContext, context, body, symbol);
             }
         }
 
-        private void Analyze(SyntaxNodeAnalysisContext context, CSharpSyntaxNode body, ISymbol symbol)
+        private void Analyze(SonarAnalysisContext analysisContext, SyntaxNodeAnalysisContext context, CSharpSyntaxNode body, ISymbol symbol)
         {
-            var enabledAnalyzers = symbolicExecutionAnalyzerFactory.GetEnabledAnalyzers(context);
+            var isTestProject = analysisContext.IsTestProject(context.Compilation, context.Options);
+            var isScannerRun = analysisContext.IsScannerRun(context.Options);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs:104]: I need these two here for the future PR. That's why I don't pass `context` itself.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Sonar/SymbolicExecutionAnalyzerFactoryTest.cs [hunk 2]
```diff
@@ -88,13 +85,22 @@ public void GetEnabledAnalyzers_ReturnsEmptyList_WhenDiagnosticsAreDisabled()
...
-                    .SelectMany(analyzer => analyzer.SupportedDiagnostics.Select(descriptor => descriptor.Id))
-                    .ToList();
+            var analyzers = sut.GetEnabledAnalyzers(context, false, false).ToList();
+            var enabledAnalyzers = analyzers.SelectMany(analyzer => analyzer.SupportedDiagnostics.Select(descriptor => descriptor.Id)).ToList();
+
+            enabledAnalyzers.Should().BeEmpty();
+        }
+
+        [TestMethod]
+        public void GetEnabledAnalyzers_ReturnsEmptyList_OnTestProjectScannerRun()  // Old SE rules do not run on test code
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Sonar/SymbolicExecutionAnalyzerFactoryTest.cs:95]: The comment is a bit misleading, as there are rules that are run in standalone nuget package case.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Sonar/SymbolicExecutionAnalyzerFactoryTest.cs [hunk 3]
```diff
@@ -88,13 +85,22 @@ public void GetEnabledAnalyzers_ReturnsEmptyList_WhenDiagnosticsAreDisabled()
...
+        {
+            var sut = new SymbolicExecutionAnalyzerFactory();
+            var diagnostics = symbolicExecutionRuleIds.ToImmutableDictionary(v => v, _ => ReportDiagnostic.Warn);
+            var context = CreateSyntaxNodeAnalysisContext(diagnostics);
+            var analyzers = sut.GetEnabledAnalyzers(context, true, true).ToList();
+            var enabledAnalyzers = analyzers.SelectMany(analyzer => analyzer.SupportedDiagnostics.Select(descriptor => descriptor.Id)).ToList();
 
-            CollectionAssert.AreEquivalent(new List<string>(), enabledAnalyzers);
+            enabledAnalyzers.Should().BeEmpty();
         }
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Sonar/SymbolicExecutionAnalyzerFactoryTest.cs:104]: Please add another test like the one above with the value of false for the isScannerRun parameter. That will check that the "All" scoped rules are run in the Nuget case.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Please check comments.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Sonar/SymbolicExecutionAnalyzerFactoryTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Sonar/SymbolicExecutionAnalyzerFactoryTest.cs:104]: The entire `SymbolicExecutionAnalyzerFactory` and `SymbolicExecutionAnalyzerFactoryTest` will be deleted quite soon. All these various scenarios will be tested in #5159

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**